compiler/defer.go

```diff
@@ -234,41 +234,88 @@ func (b *builder) createInvokeCheckpoint() {
 	continueBB := b.insertBasicBlock("")
 	b.CreateCondBr(isZero, continueBB, b.landingpad)
 	b.SetInsertPointAtEnd(continueBB)
-	b.blockExits[b.currentBlock] = continueBB
+	b.currentBlockInfo.exit = continueBB
 }
 
-// isInLoop checks if there is a path from a basic block to itself.
-func isInLoop(start *ssa.BasicBlock) bool {
-	// Use a breadth-first search to scan backwards through the block graph.
-	queue := []*ssa.BasicBlock{start}
-	checked := map[*ssa.BasicBlock]struct{}{}
-
-	for len(queue) > 0 {
-		// pop a block off of the queue
-		block := queue[len(queue)-1]
-		queue = queue[:len(queue)-1]
-
-		// Search through predecessors.
-		// Searching backwards means that this is pretty fast when the block is close to the start of the function.
-		// Defers are often placed near the start of the function.
-		for _, pred := range block.Preds {
-			if pred == start {
-				// cycle found
-				return true
-			}
+// isInLoop checks if there is a path from the current block to itself.
+// Use Tarjan's strongly connected components algorithm to search for cycles.
+// A one-node SCC is a cycle iff there is an edge from the node to itself.
+// A multi-node SCC is always a cycle.
+func (b *builder) isInLoop() bool {
+	if b.currentBlockInfo.tarjan.index == 0 {
+		b.strongConnect(b.currentBlock)
+	}
+	return b.currentBlockInfo.tarjan.cyclic
+}
 
-			if _, ok := checked[pred]; ok {
-				// block already checked
-				continue
-			}
+func (b *builder) strongConnect(block *ssa.BasicBlock) {
+	// Assign a new index.
+	assignedIndex := b.tarjanIndex + 1
+	b.tarjanIndex = assignedIndex
+
+	// Apply the new index.
+	blockIndex := block.Index
+	node := &b.blockInfo[blockIndex].tarjan
+	node.index = assignedIndex
+	node.lowLink = assignedIndex
+
+	// Push the node onto the stack.
+	node.onStack = true
+	b.tarjanStack = append(b.tarjanStack, uint(blockIndex))
+
+	// Process the successors.
+	for _, successor := range block.Succs {
+		// Look up the successor's state.
+		successorIndex := successor.Index
+		if successorIndex == blockIndex {
+			// Handle a self-cycle specially.
+			node.cyclic = true
+			continue
+		}
+		successorNode := &b.blockInfo[successorIndex].tarjan
+
+		if successorNode.index == 0 {
```

Comment:
I prefer this syntax:

```go
switch {
case successorNode.index == 0:
	// This node has not yet been visited.
	b.strongConnect(successor)
case !successorNode.onStack:
	// This node has been visited, but is in a different SCC.
	// Ignore it.
	continue
}
```

Also note spelling/grammar corrections. :smile_cat:

Reply:
Done.